frontend-react/src/utils/DiffCompare/JsonDiffer.ts

```diff
@@ -0,0 +1,328 @@
+        (acc, each: Marker): MarkerInsert[] => [
+            ...acc,
+            { pos: each.start, mark: "<mark>" },
+            { pos: each.end, mark: "</mark>" },
+        ],
+        [] as MarkerInsert[]
+    );
+
+    // we reverse sort by pos. Work must be done from back to front
+    inserts.sort((a, b) => b.pos - a.pos);
+
+    // go through and inject <mark> or </mark> at each pos
+    return inserts.reduce(
+        (acc: string, each: MarkerInsert) =>
+            `${acc.slice(0, each.pos)}${each.mark}${acc.slice(each.pos)}`,
+        jsonStr
+    );
+};
+
+/**
+ * Diff compares two jsons and returns a map of the differences. You probably want to use jsonDifferMarkup
+ * @param leftData
+ * @param rightData
+ * @return JsonDiffResult Called by jsonDifferMarkup.
+ */
+const jsonDiffer = (
+    leftData: SourceMapResult,
+    rightData: SourceMapResult
+): JsonDiffResult => {
+    // diff the keys. If the key is different, then just consider the value of that key to be different.
+    const leftKeys = Object.keys(leftData.pointers);
+    const rightKeys = Object.keys(rightData.pointers);
+
+    // this is looking for diffs between the two lists.
+    let addedLeftKeys = leftKeys.filter(
+        (key) => key.length && !rightKeys.includes(key)
+    );
+    let addedRightKeys = rightKeys.filter(
+        (key) => key.length && !leftKeys.includes(key)
+    );
+
+    // now we want intersection (aka NOT changed and see if the values have changed).
+    let intersection = leftKeys.filter((key) => rightKeys.includes(key));
+
+    // for readability improvements only, pull out start/end values
+    const getStartEnd = (key: string, data: SourceMapResult) => {
+        return [data.pointers[key].value.pos, data.pointers[key].valueEnd.pos];
+    };
+
+    // inline getValue() improves readability only. slices out the string of a given value
+    const getValue = (key: string, data: SourceMapResult): string =>
+        data.json.slice(...getStartEnd(key, data));
+
+    // we pull out the value of unchanged keys and see if that's different.
+    let changedKeys = intersection.filter(
+        (key) =>
+            key !== "" && getValue(key, leftData) !== getValue(key, rightData)
+    );
+
+    // now extract just the node leaves from the keys. This is because technically the content of each parent
+    // has changed, but we really only think about the leaf nodes as being diff
+    changedKeys = extractLeafNodes(changedKeys);
+
+    return {
+        addedLeftKeys,
+        addedRightKeys,
+        changedKeys,
+    };
+};
+
+/**
+ * Given two json find the difference, then return a normalized json string and one with <mark>s
+ * added for the differences.
+ *
+ * @param leftJson Valid Json object
+ * @param rightJson Valid Json object
+ */
+export const jsonDifferMarkup = (
+    leftJson: any,
+    rightJson: any
+): DifferMarkupResult => {
+    // left and right should be json objects, but there's really no way to typescript enforce it.
```

Comment:
will this error somewhere if strings are passed instead of pojos?

Reply:
This is a good question. A string with quotes doesn't fail because it can be passed into JSON.parse(`"This doesn't fail"`). But the function doesn't do anything. I added a test to make sure this is the case.